src/device/ping-device.cpp

```diff
@@ -0,0 +1,94 @@
+#include <ping-device.h>
+
+#include <ping-message-common.h>
+#include <ping-message.h>
+#include <ping-time.h>
+
+bool PingDevice::initialize() {
+  return request(CommonId::PROTOCOL_VERSION)
+    && request(CommonId::DEVICE_INFORMATION);
+}
+
+ping_message* PingDevice::read() {
+  char b;
+  int result = _port.read(&b, 1);
+  if (result > 0) {
+    if (_parser.parseByte(static_cast<char>(b)) == PingParser::NEW_MESSAGE) {
+      return &_parser.rxMessage;
+    }
+  }
+  return nullptr;
+}
+
+ping_message* PingDevice::request(uint16_t id, int timeoutMs) {
+  common_general_request msg;
+  msg.set_requested_id(id);
+  writeMessage(msg);
+  ping_message* reply = waitMessage(id, timeoutMs);
+  if (reply && reply->message_id() == CommonId::NACK) {
+    common_nack* m = static_cast<common_nack*>(reply);
+    if (m->nacked_id() != CommonId::GENERAL_REQUEST) {
+      return nullptr;
+    }
+  }
+
+  return reply;
+}
+
+ping_message* PingDevice::waitMessage(uint16_t id, int timeoutMs) {
+  int tstart = PingTime::timeMs();
+  while (PingTime::timeMs() < tstart + timeoutMs) {
+    // TODO make this totally nonblocking
+    // this will block up to 0.1s for data to be available
+    ping_message* msg = read();
+
+    if (!msg) {
+      continue;
+    }
+
+    _handleMessage(msg);
+
+    if (msg->message_id() == id || msg->message_id() == CommonId::NACK) {
+      return msg;
+    }
+    // sleep for 1ms to prevent cpu spinlock
+    PingTime::microsecondDelay(1000);
+  }
+
+  return nullptr;
+}
+
+int PingDevice::write(const char* data, int nBytes) { return _port.write(data, nBytes); }
+
+void PingDevice::writeMessage(ping_message& message) {
+  message.updateChecksum();
```

Comment:
Why writeMessage updates the checksum ? So writeMessage can change the message variable ?

Reply:
it is to avoid user error and provide a device api that is easier to use, yes this function is allowed to change the message variable.

I thought about taking it out, but that just forces the user to call updateChecksum, and we don't need to do that to the user.